File: backend-fastapi/models/user_model.py
"""
TaskNet - User Model (SQLAlchemy ORM)
File: backend-fastapi/models/user_model.py

Declarative mapping for the `users` table and related junction tables.
All FKs, unique constraints, and ORM relationships are explicitly defined.
"""

import enum
from datetime import date, datetime
from sqlalchemy import (
    BigInteger, Boolean, Column, Date, DateTime, Enum,
    ForeignKey, Integer, String, Table, Text, func,
)
from sqlalchemy.orm import DeclarativeBase, relationship


# ---------------------------------------------------------------------------
# Base
# ---------------------------------------------------------------------------
class Base(DeclarativeBase):
    pass


# ---------------------------------------------------------------------------
# Enums
# ---------------------------------------------------------------------------
class UserRole(str, enum.Enum):
    PRINCIPAL   = "principal"
    DEAN        = "dean"
    COORDINATOR = "coordinator"
    REGISTRAR   = "registrar"
    TEACHER     = "teacher"


class UserStatus(str, enum.Enum):
    ACTIVE      = "active"
    DEACTIVATED = "deactivated"


# ---------------------------------------------------------------------------
# Junction: user ↔ grade_level  (many-to-many)
# ---------------------------------------------------------------------------
user_grade_levels = Table(
    "user_grade_levels",
    Base.metadata,
    Column("user_id",        Integer, ForeignKey("users.id", ondelete="CASCADE"), primary_key=True),
    Column("grade_level_id", Integer, ForeignKey("grade_levels.id", ondelete="CASCADE"), primary_key=True),
)

# Junction: user ↔ subject  (many-to-many via subject_grade_assignments)
# A teacher can teach multiple subject-grade pairs.
user_subject_grade = Table(
    "user_subject_grade_assignments",
    Base.metadata,
    Column("id",             Integer, primary_key=True, autoincrement=True),
    Column("user_id",        Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False),
    Column("grade_level_id", Integer, ForeignKey("grade_levels.id", ondelete="CASCADE"), nullable=False),
    Column("subject_id",     Integer, ForeignKey("subjects.id",      ondelete="CASCADE"), nullable=False),
)


# ---------------------------------------------------------------------------
# GradeLevel
# ---------------------------------------------------------------------------
class GradeLevel(Base):
    __tablename__ = "grade_levels"

    id    = Column(Integer, primary_key=True, autoincrement=True)
    name  = Column(String(50), nullable=False, unique=True)   # e.g. "Grade 1"

    # Relationships
    users = relationship("User", secondary=user_grade_levels, back_populates="grade_levels")


# ---------------------------------------------------------------------------
# Subject
# ---------------------------------------------------------------------------
class Subject(Base):
    __tablename__ = "subjects"

    id   = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String(100), nullable=False, unique=True)   # e.g. "Mathematics"


# ---------------------------------------------------------------------------
# Department  (used by Dean)
# ---------------------------------------------------------------------------
class Department(Base):
    __tablename__ = "departments"

    id          = Column(Integer, primary_key=True, autoincrement=True)
    name        = Column(String(150), nullable=False, unique=True)
    grade_range = Column(String(50))

    deans = relationship("User", back_populates="department")


# ---------------------------------------------------------------------------
# CoordinatorType  (used by Coordinator)
# ---------------------------------------------------------------------------
class CoordinatorType(Base):
    __tablename__ = "coordinator_types"

    id   = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String(100), nullable=False, unique=True)

    coordinators = relationship("User", back_populates="coordinator_type")


# ---------------------------------------------------------------------------
# User  (central entity – represents all staff roles)
# ---------------------------------------------------------------------------
class User(Base):
    __tablename__ = "users"

    # ── Primary Key ─────────────────────────────────────────────────────────
    id = Column(Integer, primary_key=True, autoincrement=True)

    # ── Identity ─────────────────────────────────────────────────────────────
    employee_no = Column(String(20), unique=True, nullable=True)   # displayed as "ID" in UI
    first_name  = Column(String(80),  nullable=False)
    middle_name = Column(String(80),  nullable=True)
    last_name   = Column(String(80),  nullable=False)
    suffix      = Column(String(20),  nullable=True)

    # ── Account ──────────────────────────────────────────────────────────────
    username        = Column(String(80), unique=True, nullable=True)
    hashed_password = Column(String(256), nullable=False)
    email           = Column(String(200), unique=True, nullable=False)

    # ── Role & Status ─────────────────────────────────────────────────────────
    role   = Column(Enum(UserRole,   name="userrole",   create_type=False, values_callable=lambda x: [e.value for e in x]), nullable=False, default=UserRole.TEACHER)
    status = Column(Enum(UserStatus, name="userstatus", create_type=False, values_callable=lambda x: [e.value for e in x]), nullable=False, default=UserStatus.ACTIVE)

    # ── Contact ──────────────────────────────────────────────────────────────
    contact_number = Column(String(30), nullable=True)
    birthdate      = Column(Date,       nullable=True)
    address        = Column(Text,       nullable=True)

    # ── Government IDs ────────────────────────────────────────────────────────
    tin_number    = Column(String(30), nullable=True)
    gsis_number   = Column(String(30), nullable=True)
    pagibig_number = Column(String(30), nullable=True)
    philhealth_number = Column(String(30), nullable=True)

    # ── Employment ───────────────────────────────────────────────────────────
    date_hired       = Column(Date, nullable=True)
    date_of_appointment = Column(Date, nullable=True)

    # ── Profile Photo ─────────────────────────────────────────────────────────
    avatar_url = Column(String(500), nullable=True)

    # ── FK: Department (for Deans) ────────────────────────────────────────────
    department_id = Column(Integer, ForeignKey("departments.id", ondelete="SET NULL"), nullable=True)

    # ── FK: CoordinatorType (for Coordinators) ────────────────────────────────
    coordinator_type_id = Column(Integer, ForeignKey("coordinator_types.id", ondelete="SET NULL"), nullable=True)

    # ── First login flag (cleared after user completes profile setup) ─────────
    first_login = Column(Boolean, nullable=False, default=True, server_default='true')

    # ── Audit ─────────────────────────────────────────────────────────────────
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)
    # Granular last-updated timestamps for UI display
    personal_info_updated_at      = Column(DateTime(timezone=True), nullable=True)
    academic_delegation_updated_at = Column(DateTime(timezone=True), nullable=True)

    # ── Relationships ─────────────────────────────────────────────────────────
    department       = relationship("Department",      back_populates="deans")
    coordinator_type = relationship("CoordinatorType", back_populates="coordinators")
    grade_levels     = relationship("GradeLevel", secondary=user_grade_levels, back_populates="users")
    delegation_history = relationship("UserDelegationHistory", back_populates="user",
                                      foreign_keys="UserDelegationHistory.user_id",
                                      order_by="UserDelegationHistory.changed_at.desc()")

    # ── Computed helpers ──────────────────────────────────────────────────────
    @property
    def full_name(self) -> str:
        parts = [self.first_name]
        if self.middle_name:
            parts.append(self.middle_name[0] + ".")
        parts.append(self.last_name)
        if self.suffix:
            parts.append(self.suffix)
        return " ".join(parts)

    @property
    def is_active(self) -> bool:
        return self.status == UserStatus.ACTIVE


# ---------------------------------------------------------------------------
# Delegation History  (auto-recorded whenever academic delegation changes)
# ---------------------------------------------------------------------------
class UserDelegationHistory(Base):
    __tablename__ = "user_delegation_history"

    id             = Column(Integer,  primary_key=True, autoincrement=True)
    user_id        = Column(Integer,  ForeignKey("users.id", ondelete="CASCADE"),   nullable=False)
    changed_by_id  = Column(Integer,  ForeignKey("users.id", ondelete="SET NULL"),  nullable=True)
    changed_at     = Column(DateTime(timezone=True), server_default=func.now(),     nullable=False)

    # Snapshot of the delegation state at the time of change
    role                   = Column(String(20),  nullable=True)
    grade_level_handled    = Column(String(50),  nullable=True)   # Dean: grade level name
    coordinator_type       = Column(String(100), nullable=True)   # Coordinator: type name
    subject_grade_summary  = Column(Text,        nullable=True)   # "Math (Grade 1), ..."
    notes                  = Column(String(200), nullable=True)   # Human-readable summary

    # Relationships
    user       = relationship("User", back_populates="delegation_history", foreign_keys=[user_id])
    changed_by = relationship("User", foreign_keys=[changed_by_id])

File: frontend-flutter/lib/models/user_model.dart
// TaskNet - User Dart Models
// File: frontend-flutter/lib/models/user_model.dart

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------
enum UserRole {
  principal,
  dean,
  coordinator,
  registrar,
  teacher;

  String get displayName => switch (this) {
        UserRole.principal => 'Principal',
        UserRole.dean => 'Dean',
        UserRole.coordinator => 'Coordinator',
        UserRole.registrar => 'Registrar',
        UserRole.teacher => 'Teacher',
      };
}

enum UserStatus {
  active,
  deactivated;

  String get displayName => switch (this) {
        UserStatus.active => 'Active',
        UserStatus.deactivated => 'Deactivated',
      };
}

// ---------------------------------------------------------------------------
// Supporting types
// ---------------------------------------------------------------------------
class GradeLevelModel {
  final int id;
  final String name;

  const GradeLevelModel({required this.id, required this.name});

  factory GradeLevelModel.fromJson(Map<String, dynamic> j) =>
      GradeLevelModel(id: j['id'] as int, name: j['name'] as String);

  Map<String, dynamic> toJson() => {'id': id, 'name': name};
}

class SubjectModel {
  final int id;
  final String name;

  const SubjectModel({required this.id, required this.name});

  factory SubjectModel.fromJson(Map<String, dynamic> j) =>
      SubjectModel(id: j['id'] as int, name: j['name'] as String);
}

class DepartmentModel {
  final int id;
  final String name;
  final String? gradeRange;

  const DepartmentModel({required this.id, required this.name, this.gradeRange});

  factory DepartmentModel.fromJson(Map<String, dynamic> j) => DepartmentModel(
        id: j['id'] as int,
        name: j['name'] as String,
        gradeRange: j['grade_range'] as String?,
      );
}

class CoordinatorTypeModel {
  final int id;
  final String name;

  const CoordinatorTypeModel({required this.id, required this.name});

  factory CoordinatorTypeModel.fromJson(Map<String, dynamic> j) =>
      CoordinatorTypeModel(id: j['id'] as int, name: j['name'] as String);
}

class SubjectGradeAssignment {
  final int gradeLevelId;
  final int subjectId;

  const SubjectGradeAssignment({required this.gradeLevelId, required this.subjectId});

  factory SubjectGradeAssignment.fromJson(Map<String, dynamic> j) => SubjectGradeAssignment(
        gradeLevelId: j['grade_level_id'] as int,
        subjectId: j['subject_id'] as int,
      );

  Map<String, dynamic> toJson() => {
        'grade_level_id': gradeLevelId,
        'subject_id': subjectId,
      };
}

// ---------------------------------------------------------------------------
// User Brief (table row)
// ---------------------------------------------------------------------------
class UserBriefModel {
  final int id;
  final String? employeeNo;
  final String fullName;
  final String? username;
  final String? contactNumber;
  final String email;
  final UserRole role;
  final UserStatus status;
  final List<GradeLevelModel> gradeLevels;
  final List<String> subjects;

  const UserBriefModel({
    required this.id,
    this.employeeNo,
    required this.fullName,
    this.username,
    this.contactNumber,
    required this.email,
    required this.role,
    required this.status,
    required this.gradeLevels,
    required this.subjects,
  });

  factory UserBriefModel.fromJson(Map<String, dynamic> j) => UserBriefModel(
        id: j['id'] as int,
        employeeNo: j['employee_no'] as String?,
        fullName: j['full_name'] as String,
        username: j['username'] as String?,
        contactNumber: j['contact_number'] as String?,
        email: j['email'] as String,
        role: UserRole.values.byName(j['role'] as String),
        status: UserStatus.values.byName(j['status'] as String),
        gradeLevels: (j['grade_levels'] as List)
            .map((e) => GradeLevelModel.fromJson(e as Map<String, dynamic>))
            .toList(),
        subjects: (j['subjects'] as List).map((e) => e as String).toList(),
      );

  String get gradeLevelsDisplay =>
      gradeLevels.isEmpty ? '—' : gradeLevels.map((g) => g.name).join(', ');

  String get subjectsDisplay =>
      subjects.isEmpty ? '—' : subjects.join(', ');
}

// ---------------------------------------------------------------------------
// User Detail (full record)
// ---------------------------------------------------------------------------
class UserDetailModel {
  final int id;
  final String? employeeNo;
  final String firstName;
  final String? middleName;
  final String lastName;
  final String? suffix;
  final String fullName;
  final String? username;
  final String email;
  final UserRole role;
  final UserStatus status;

  final String? contactNumber;
  final String? birthdate;
  final String? address;

  final String? tinNumber;
  final String? gsisNumber;
  final String? pagibigNumber;
  final String? philhealthNumber;

  final String? dateHired;
  final String? dateOfAppointment;

  final String? avatarUrl;
  final DepartmentModel? department;
  final CoordinatorTypeModel? coordinatorType;
  final List<GradeLevelModel> gradeLevels;
  final List<SubjectGradeAssignment> subjectGradeAssignments;
  final bool firstLogin;
  final DateTime? personalInfoUpdatedAt;
  final DateTime? academicDelegationUpdatedAt;

  const UserDetailModel({
    required this.id,
    this.employeeNo,
    required this.firstName,
    this.middleName,
    required this.lastName,
    this.suffix,
    required this.fullName,
    this.username,
    required this.email,
    required this.role,
    required this.status,
    this.contactNumber,
    this.birthdate,
    this.address,
    this.tinNumber,
    this.gsisNumber,
    this.pagibigNumber,
    this.philhealthNumber,
    this.dateHired,
    this.dateOfAppointment,
    this.avatarUrl,
    this.department,
    this.coordinatorType,
    required this.gradeLevels,
    this.subjectGradeAssignments = const [],
    this.firstLogin = false,
    this.personalInfoUpdatedAt,
    this.academicDelegationUpdatedAt,
  });

  factory UserDetailModel.fromJson(Map<String, dynamic> j) => UserDetailModel(
        id: j['id'] as int,
        employeeNo: j['employee_no'] as String?,
        firstName: j['first_name'] as String,
        middleName: j['middle_name'] as String?,
        lastName: j['last_name'] as String,
        suffix: j['suffix'] as String?,
        fullName: j['full_name'] as String,
        username: j['username'] as String?,
        email: j['email'] as String,
        role: UserRole.values.byName(j['role'] as String),
        status: UserStatus.values.byName(j['status'] as String),
        contactNumber: j['contact_number'] as String?,
        birthdate: j['birthdate'] as String?,
        address: j['address'] as String?,
        tinNumber: j['tin_number'] as String?,
        gsisNumber: j['gsis_number'] as String?,
        pagibigNumber: j['pagibig_number'] as String?,
        philhealthNumber: j['philhealth_number'] as String?,
        dateHired: j['date_hired'] as String?,
        dateOfAppointment: j['date_of_appointment'] as String?,
        avatarUrl: j['avatar_url'] as String?,
        department: j['department'] == null
            ? null
            : DepartmentModel.fromJson(j['department'] as Map<String, dynamic>),
        coordinatorType: j['coordinator_type'] == null
            ? null
            : CoordinatorTypeModel.fromJson(j['coordinator_type'] as Map<String, dynamic>),
        gradeLevels: (j['grade_levels'] as List)
            .map((e) => GradeLevelModel.fromJson(e as Map<String, dynamic>))
            .toList(),
        subjectGradeAssignments: ((j['subject_grade_assignments'] ?? []) as List)
            .map((e) => SubjectGradeAssignment.fromJson(e as Map<String, dynamic>))
            .toList(),
        firstLogin: j['first_login'] as bool? ?? false,
        personalInfoUpdatedAt:
            j['personal_info_updated_at'] == null ? null : DateTime.parse(j['personal_info_updated_at'] as String),
        academicDelegationUpdatedAt:
            j['academic_delegation_updated_at'] == null ? null : DateTime.parse(j['academic_delegation_updated_at'] as String),
      );
}

// ---------------------------------------------------------------------------
// Delegation History
// ---------------------------------------------------------------------------
class DelegationHistoryItem {
  final int id;
  final DateTime changedAt;
  final String? changedByName;
  final String? role;
  final String? gradeLevelHandled;
  final String? coordinatorType;
  final String? subjectGradeSummary;
  final String? notes;

  const DelegationHistoryItem({
    required this.id,
    required this.changedAt,
    this.changedByName,
    this.role,
    this.gradeLevelHandled,
    this.coordinatorType,
    this.subjectGradeSummary,
    this.notes,
  });

  factory DelegationHistoryItem.fromJson(Map<String, dynamic> j) => DelegationHistoryItem(
        id: j['id'] as int,
        changedAt: DateTime.parse(j['changed_at'] as String),
        changedByName: j['changed_by_name'] as String?,
        role: j['role'] as String?,
        gradeLevelHandled: j['grade_level_handled'] as String?,
        coordinatorType: j['coordinator_type'] as String?,
        subjectGradeSummary: j['subject_grade_summary'] as String?,
        notes: j['notes'] as String?,
      );
}

// ---------------------------------------------------------------------------
// List response wrapper
// ---------------------------------------------------------------------------
class UserListResponse {
  final List<UserBriefModel> active;
  final List<UserBriefModel> deactivated;
  final int total;

  const UserListResponse({
    required this.active,
    required this.deactivated,
    required this.total,
  });

  factory UserListResponse.fromJson(Map<String, dynamic> j) => UserListResponse(
        active: (j['active'] as List)
            .map((e) => UserBriefModel.fromJson(e as Map<String, dynamic>))
            .toList(),
        deactivated: (j['deactivated'] as List)
            .map((e) => UserBriefModel.fromJson(e as Map<String, dynamic>))
            .toList(),
        total: j['total'] as int,
      );
}
